src/core/expressionvariablemodel.h

```diff
@@ -17,40 +17,65 @@
 #define EXPRESSIONVARIABLEMODEL_H
 
 #include <QStandardItemModel>
+#include <qgsproject.h>
 
 class ExpressionVariableModel : public QStandardItemModel
 {
     Q_OBJECT
 
+    Q_PROPERTY( QgsProject *currentProject READ currentProject WRITE setCurrentProject NOTIFY currentProjectChanged )
+
   public:
     enum Roles
     {
       VariableName = Qt::UserRole,
-      VariableValue
+      VariableValue,
+      VariableScopeRole,
+      VariableEditable = Qt::EditRole
+    };
+
+    enum class VariableScope
+    {
+      ApplicationScope,
+      ProjectScope
     };
 
     explicit ExpressionVariableModel( QObject *parent = nullptr );
 
     bool setData( const QModelIndex &index, const QVariant &value, int role ) override;
 
-    Q_INVOKABLE void addCustomVariable( const QString &varName, const QString &varVal );
+    Q_INVOKABLE void addCustomVariable( const QString &varName, const QString &varVal, const int &rowIndex = -1 );
 
     Q_INVOKABLE void removeCustomVariable( int row );
```

Comment:
I think I'd change this to allow for removing of variable via name + scope instead of a slightly more obscure row index.

```suggestion
    Q_INVOKABLE void removeVariable( VariableScope scope, const QString &name );
```